### ISSUE TITLE
chat: input_ref.current.focus is not a function errors

### ISSUE THREAD

**Issue by haraldschilly** [sagemathinc/cocalc#3975]
two cases of `r.input_ref.current.focus is not a function`

there is also https://github.com/sagemathinc/cocalc/issues/3849 fixing the very same error (probably different code)

```
TypeError: r.input_ref.current.focus is not a function
    at r.on_preview_button_click (https://cocalc.com/static/smc-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:5:1498373)
    at Object.<anonymous> (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4461316)
    at f (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4461354)
    at https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4461992
    at S (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4462081)
    at E (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4462494)
    at M (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4462329)
    at L (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4463416)
    at Cn (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4490793)
    at Ls (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4558906)
```

and

```
TypeError: this.input_ref.current.focus is not a function
    at Object.send_chat (https://cocalc.com/static/smc-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:5:3132436)
    at Object.on_send_click (https://cocalc.com/static/smc-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:5:3132145)
    at Object.<anonymous> (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4461316)
    at f (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4461354)
    at https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4461992
    at S (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4462081)
    at E (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4462517)
    at M (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4462329)
    at L (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4463416)
    at Cn (https://cocalc.com/static/0-476d15d6d25b7b4ef719.cacheme.js?476d15d6d25b7b4ef719:6:4490793)
```

**Comment by williamstein** [comment]
I just happened to hit this immediately the first second I started looking into rewriting chat to use react-virtualized!  The problem is made clear by NOT being lazy with typescript in smc_chat.tsx.  This one change:
```
-  private input_ref: any;
+  private input_ref = React.createRef<HTMLTextAreaElement>();
 
   constructor(props: ChatRoomProps, context: any) {
     super(props, context);
     this.state = { preview: "" };
-    this.input_ref = React.createRef<HTMLTextAreaElement>();
   }
```
(see [here](https://medium.com/@martin_hotell/react-refs-with-typescript-a32d56c4d315)), makes it so typescript gets pissed about all unchecked uses of input_ref.  That previous `private input_ref: any;` that @johnjeng had written [when fixing image paste and mentions in chat](https://github.com/sagemathinc/cocalc/commit/3b5e656e69d6cd56998ff24c27a6cc985dbfc989) was worse than not having an any there at all, because it hid all the typescript errors that would protect against tracebacks like this.

I'll just fix this in my soon-to-be-finished react-virtualized branch...

**Comment by williamstein** [comment]
Oh, and I was doing similar bad things with React.createRef in my code, and am going to fix it.

**Comment by haraldschilly** [comment]
This should be live, right? Because I get these errors: (the commit is 4 days old)

smc_git_rev: 2dcecb1826eb6278da24270ed2533f9edded2ec3

Date: 2019-08-09 16:16:56.204977

`r.input_ref.current.focus is not a function`

```
TypeError: r.input_ref.current.focus is not a function
    at r.on_preview_button_click (https://cocalc.com/static/smc-06b535768c5e1c29d414.cacheme.js?06b535768c5e1c29d414:5:1023089)
    at Object.<anonymous> (https://cocalc.com/static/0-06b535768c5e1c29d414.cacheme.js?06b535768c5e1c29d414:6:4465284)
    at f (https://cocalc.com/static/0-06b535768c5e1c29d414.cacheme.js?06b535768c5e1c29d414:6:4465322)
    at https://cocalc.com/static/0-06b535768c5e1c29d414.cacheme.js?06b535768c5e1c29d414:6:4465960
    at k (https://cocalc.com/static/0-06b535768c5e1c29d414.cacheme.js?06b535768c5e1c29d414:6:4466049)
    at E (https://cocalc.com/static/0-06b535768c5e1c29d414.cacheme.js?06b535768c5e1c29d414:6:4466462)
    at M (https://cocalc.com/static/0-06b535768c5e1c29d414.cacheme.js?06b535768c5e1c29d414:6:4466297)
    at L (https://cocalc.com/static/0-06b535768c5e1c29d414.cacheme.js?06b535768c5e1c29d414:6:4467384)
    at xr (https://cocalc.com/static/0-06b535768c5e1c29d414.cacheme.js?06b535768c5e1c29d414:6:4494761)
    at Ls (https://cocalc.com/static/0-06b535768c5e1c29d414.cacheme.js?06b535768c5e1c29d414:6:4562874)
```

---

chrome: `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.100 Safari/537.36`

smc_git_rev: 3420b9cebfcc3b14830d416648d75c7818823d57

Date: 2019-08-09 20:54:20.652119

`this.input_ref.current.focus is not a function`

```
TypeError: this.input_ref.current.focus is not a function
    at Object.send_chat (https://cocalc.com/static/smc-354d7b81dddbaa1d3d64.cacheme.js?354d7b81dddbaa1d3d64:5:3147851)
    at Object.on_send_click (https://cocalc.com/static/smc-354d7b81dddbaa1d3d64.cacheme.js?354d7b81dddbaa1d3d64:5:3147577)
    at Object.<anonymous> (https://cocalc.com/static/0-354d7b81dddbaa1d3d64.cacheme.js?354d7b81dddbaa1d3d64:6:4465284)
    at f (https://cocalc.com/static/0-354d7b81dddbaa1d3d64.cacheme.js?354d7b81dddbaa1d3d64:6:4465322)
    at https://cocalc.com/static/0-354d7b81dddbaa1d3d64.cacheme.js?354d7b81dddbaa1d3d64:6:4465960
    at k (https://cocalc.com/static/0-354d7b81dddbaa1d3d64.cacheme.js?354d7b81dddbaa1d3d64:6:4466049)
    at E (https://cocalc.com/static/0-354d7b81dddbaa1d3d64.cacheme.js?354d7b81dddbaa1d3d64:6:4466485)
    at M (https://cocalc.com/static/0-354d7b81dddbaa1d3d64.cacheme.js?354d7b81dddbaa1d3d64:6:4466297)
    at L (https://cocalc.com/static/0-354d7b81dddbaa1d3d64.cacheme.js?354d7b81dddbaa1d3d64:6:4467384)
    at xr (https://cocalc.com/static/0-354d7b81dddbaa1d3d64.cacheme.js?354d7b81dddbaa1d3d64:6:4494761)
```

**Comment by williamstein** [comment]
That was in some old coffeescript, where my declarations in typescript mentioned above didn't have any impact. I'll fix.

**Comment by haraldschilly** [comment]
I'm aware of your fix in db79344, which looks good, but … now there is a complaint that `focus` isn't a function?

chrome: `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 Safari/537.36`

smc_git_rev: 3c2f618cda977193f076f299c61c43d4dfc96bf2

Date: 2019-08-29 13:06:23.471464

`this.input_ref.current.focus is not a function`

```
TypeError: this.input_ref.current.focus is not a function
    at Object.send_chat (https://cocalc.com/static/smc-36be0222b27b58098e8c.cacheme.js?36be0222b27b58098e8c:5:3180830)
    at Object.on_send_click (https://cocalc.com/static/smc-36be0222b27b58098e8c.cacheme.js?36be0222b27b58098e8c:5:3180524)
    at Object.<anonymous> (https://cocalc.com/static/0-36be0222b27b58098e8c.cacheme.js?36be0222b27b58098e8c:6:4486705)
    at f (https://cocalc.com/static/0-36be0222b27b58098e8c.cacheme.js?36be0222b27b58098e8c:6:4486743)
    at https://cocalc.com/static/0-36be0222b27b58098e8c.cacheme.js?36be0222b27b58098e8c:6:4487381
    at k (https://cocalc.com/static/0-36be0222b27b58098e8c.cacheme.js?36be0222b27b58098e8c:6:4487470)
    at E (https://cocalc.com/static/0-36be0222b27b58098e8c.cacheme.js?36be0222b27b58098e8c:6:4487906)
    at M (https://cocalc.com/static/0-36be0222b27b58098e8c.cacheme.js?36be0222b27b58098e8c:6:4487718)
    at L (https://cocalc.com/static/0-36be0222b27b58098e8c.cacheme.js?36be0222b27b58098e8c:6:4488805)
    at xr (https://cocalc.com/static/0-36be0222b27b58098e8c.cacheme.js?36be0222b27b58098e8c:6:4516182)
```

**Comment by williamstein** [comment]
relevant code long gone...
